File: 2023/day01_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return sum(first_digit(line) * 10 + first_digit(line, True) for line in content)


def first_digit(line: str, backward=False) -> int:
    return int([ch for ch in replace(line, backward) if ch in "1234567890"][0])


def replace(line: str, backward=False) -> str:
    if backward:
        return (
            replace(
                line.replace("twone", "1")
                .replace("zerone", "1")
                .replace("eightwo", "2")
                .replace("eighthree", "3")
                .replace("oneight", "8")
                .replace("nineight", "8"),
                False,
            )
        )[::-1]
    return (
        line.replace("twone", "2")
        .replace("zerone", "0")
        .replace("eightwo", "8")
        .replace("eighthree", "8")
        .replace("oneight", "1")
        .replace("nineight", "9")
        .replace("one", "1")
        .replace("two", "2")
        .replace("three", "3")
        .replace("four", "4")
        .replace("five", "5")
        .replace("six", "6")
        .replace("seven", "7")
        .replace("eight", "8")
        .replace("nine", "9")
        .replace("zero", "0")
    )


def main() -> None:
    with open("day01.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2023/day01-2.ts
function solution(content: string) {
  return content
    .split("\n")
    .map((line) => firstDigit(line, false) * 10 + firstDigit(line, true))
    .reduce((acc, cur) => acc + cur, 0);
}

function firstDigit(line: string, backward: boolean) {
  return parseInt(
    replace(line, backward)
      .split("")
      .find((ch) => "1234567890".includes(ch)) ?? "-1"
  );
}

function replace(line: string, backward: boolean) {
  if (backward) {
    return line
      .replace("twone", "1")
      .replaceAll("zerone", "1")
      .replaceAll("eightwo", "2")
      .replaceAll("eighthree", "3")
      .replaceAll("oneight", "8")
      .replaceAll("nineight", "8")
      .replaceAll("one", "1")
      .replaceAll("two", "2")
      .replaceAll("three", "3")
      .replaceAll("four", "4")
      .replaceAll("five", "5")
      .replaceAll("six", "6")
      .replaceAll("seven", "7")
      .replaceAll("eight", "8")
      .replaceAll("nine", "9")
      .replaceAll("zero", "0")
      .split("")
      .reverse()
      .join("");
  } else {
    return line
      .replaceAll("twone", "2")
      .replaceAll("zerone", "0")
      .replaceAll("eightwo", "8")
      .replaceAll("eighthree", "8")
      .replaceAll("oneight", "1")
      .replaceAll("nineight", "9")
      .replaceAll("one", "1")
      .replaceAll("two", "2")
      .replaceAll("three", "3")
      .replaceAll("four", "4")
      .replaceAll("five", "5")
      .replaceAll("six", "6")
      .replaceAll("seven", "7")
      .replaceAll("eight", "8")
      .replaceAll("nine", "9")
      .replaceAll("zero", "0");
  }
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
